src/external/bincode_support.rs

```diff
@@ -0,0 +1,259 @@
+
+use crate::{
+    convert::TryFrom,
+    fmt::{Braced, Hyphenated, Simple, Urn},
+    non_nil::NonNilUuid,
+    Uuid,
+};
+use bincode::{
+    config::{Config, IntEncoding},
+    de::{BorrowDecoder, Decoder},
+    enc::Encoder,
+    error::{DecodeError, EncodeError},
+    BorrowDecode, Decode, Encode,
+};
+use std::string::ToString;
+
+impl Encode for Uuid {
+    fn encode<E: Encoder>(&self, encoder: &mut E) -> Result<(), EncodeError> {
+        if encoder.config().int_encoding() == IntEncoding::Fixed {
+            usize::encode(&16, encoder)?;
+        } else {
+            u8::encode(&16, encoder)?;
+        }
+
+        <&[u8; 16]>::encode(&self.as_bytes(), encoder)?;
+
+        Ok(())
+    }
+}
+
+impl Encode for NonNilUuid {
+    fn encode<E: Encoder>(&self, encoder: &mut E) -> Result<(), EncodeError> {
+        self.get().encode(encoder)
+    }
+}
+
+impl Encode for Hyphenated {
+    fn encode<E: Encoder>(&self, encoder: &mut E) -> Result<(), EncodeError> {
+        self.as_uuid().encode(encoder)
+    }
+}
+
+impl Encode for Simple {
+    fn encode<E: Encoder>(&self, encoder: &mut E) -> Result<(), EncodeError> {
+        self.as_uuid().encode(encoder)
+    }
+}
+
+impl Encode for Urn {
+    fn encode<E: Encoder>(&self, encoder: &mut E) -> Result<(), EncodeError> {
+        self.as_uuid().encode(encoder)
+    }
+}
+
+impl Encode for Braced {
+    fn encode<E: Encoder>(&self, encoder: &mut E) -> Result<(), EncodeError> {
+        self.as_uuid().encode(encoder)
+    }
+}
+
+impl<Context> Decode<Context> for Uuid {
+    fn decode<D: Decoder<Context = Context>>(decoder: &mut D) -> Result<Self, DecodeError> {
+        if decoder.config().int_encoding() == IntEncoding::Fixed {
+            usize::decode(decoder)?;
+        } else {
+            u8::decode(decoder)?;
+        };
+
+        Ok(Uuid::from_bytes(Decode::decode(decoder)?))
+    }
+}
+impl<'de, Context> BorrowDecode<'de, Context> for Uuid {
+    fn borrow_decode<D: BorrowDecoder<'de, Context = Context>>(
+        decoder: &mut D,
+    ) -> Result<Self, DecodeError> {
+        if decoder.config().int_encoding() == IntEncoding::Fixed {
+            usize::borrow_decode(decoder)?;
+        } else {
+            u8::borrow_decode(decoder)?;
+        };
+
+        Ok(Uuid::from_bytes(BorrowDecode::borrow_decode(decoder)?))
+    }
+}
+
+impl<Context> Decode<Context> for NonNilUuid {
+    fn decode<D: Decoder<Context = Context>>(decoder: &mut D) -> Result<Self, DecodeError> {
+        let uuid = Uuid::decode(decoder)?;
+        NonNilUuid::try_from(uuid).map_err(|e| DecodeError::OtherString(e.to_string()))
+    }
+}
+impl<'de, Context> BorrowDecode<'de, Context> for NonNilUuid {
+    fn borrow_decode<D: BorrowDecoder<'de, Context = Context>>(
+        decoder: &mut D,
+    ) -> Result<Self, DecodeError> {
+        let uuid = Uuid::borrow_decode(decoder)?;
+        NonNilUuid::try_from(uuid).map_err(|e| DecodeError::OtherString(e.to_string()))
+    }
+}
+
+#[cfg(test)]
+mod bincode_tests {
```

Comment:
Besides testing the roundtrip of encode->decode, we should probably also test the output binary representation of encode. Just to ensure we know what to expect from a `Uuid` and its variants (`Simple`, etc)